File: app/src/main/java/com/kotoba/app/engine/IngatLupaEngine.java
package com.kotoba.app.engine;

import com.kotoba.app.data.model.CycleItem;
import com.kotoba.app.data.model.LearningCycle;
import com.kotoba.app.data.model.LearningObject;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Ingat/Lupa Active-Recall Memory Loop Engine.
 * Implements short-term recovery loops for Kotoba, Kanji, Verbs, and Adjectives.
 * Invariant: Any item marked LUPA is buffered into a recovery pool that must be repeated
 * strictly until LUPA = 0 before the session can complete.
 */
public class IngatLupaEngine {
    private final LearningCycle mCycleConfig;
    private final List<LearningObject> mInitialPool;

    private List<LearningObject> mCurrentPassItems;
    private final List<LearningObject> mCurrentLupaList;
    private final List<LearningObject> mAllResolvedItems;
    private final List<CycleItem> mRecordedCycleItems;

    private int mCurrentIndex;
    private int mCycleIteration;
    private boolean mIsComplete;
    private int mTotalIngatCount;
    private int mTotalLupaCount;

    public IngatLupaEngine(LearningCycle config, List<LearningObject> pool) {
        if (config == null) {
            throw new IllegalArgumentException("LearningCycle config cannot be null");
        }
        this.mCycleConfig = config;
        this.mInitialPool = pool != null ? new ArrayList<>(pool) : new ArrayList<>();

        this.mCurrentPassItems = new ArrayList<>(mInitialPool);
        this.mCurrentLupaList = new ArrayList<>();
        this.mAllResolvedItems = new ArrayList<>();
        this.mRecordedCycleItems = new ArrayList<>();

        this.mCurrentIndex = 0;
        this.mCycleIteration = 1;
        this.mTotalIngatCount = 0;
        this.mTotalLupaCount = 0;
        this.mIsComplete = mInitialPool.isEmpty();
    }

    public LearningCycle getCycleConfig() {
        return mCycleConfig;
    }

    public boolean isComplete() {
        return mIsComplete;
    }

    public boolean isRecoveryRound() {
        return mCycleIteration > 1;
    }

    public int getCycleIteration() {
        return mCycleIteration;
    }

    public int getCurrentItemIndex() {
        return mCurrentIndex;
    }

    /**
     * Navigates backwards within the current pass.
     * Only moves the pointer — does NOT undo any committed Ingat/Lupa judgment.
     * Safe to call only with index >= 0 and < getCurrentPassTotal().
     */
    public void setCurrentItemIndex(int index) {
        if (index >= 0 && index < mCurrentPassItems.size()) {
            mCurrentIndex = index;
        }
    }

    public int getCurrentPassTotal() {
        return mCurrentPassItems.size();
    }

    public int getInitialPoolSize() {
        return mInitialPool.size();
    }

    public int getTotalIngatCount() {
        return mTotalIngatCount;
    }

    public int getTotalLupaCount() {
        return mTotalLupaCount;
    }

    public int getPendingLupaCount() {
        return mCurrentLupaList.size();
    }

    public List<CycleItem> getRecordedCycleItems() {
        return Collections.unmodifiableList(mRecordedCycleItems);
    }

    public LearningObject getCurrentItem() {
        if (mIsComplete || mCurrentIndex >= mCurrentPassItems.size()) {
            return null;
        }
        return mCurrentPassItems.get(mCurrentIndex);
    }

    /**
     * Mark the current item as remembered (INGAT).
     * @param responseTimeMs response latency in milliseconds
     * @return true if session completed on this answer, false if continuing
     */
    public boolean markIngat(int responseTimeMs) {
        if (mIsComplete) return true;
        LearningObject item = getCurrentItem();
        if (item == null) return true;

        mTotalIngatCount++;
        CycleItem.PoolType poolType = isRecoveryRound() ? CycleItem.PoolType.RECOVERY : CycleItem.PoolType.NORMAL;
        CycleItem cycleItem = new CycleItem(
                mCycleConfig.getCycleId(),
                item.getId(),
                poolType,
                mCycleIteration,
                CycleItem.Response.INGAT,
                responseTimeMs,
                1
        );
        mRecordedCycleItems.add(cycleItem);

        if (!mAllResolvedItems.contains(item)) {
            mAllResolvedItems.add(item);
        }

        mCurrentIndex++;
        checkEndOfPass();
        return mIsComplete;
    }

    /**
     * Mark the current item as forgotten (LUPA).
     * Buffers the item into mCurrentLupaList for the recovery loop.
     * @param responseTimeMs response latency in milliseconds
     * @return false (always requires at least one recovery round if Lupa was chosen)
     */
    public boolean markLupa(int responseTimeMs) {
        if (mIsComplete) return true;
        LearningObject item = getCurrentItem();
        if (item == null) return true;

        mTotalLupaCount++;
        CycleItem.PoolType poolType = isRecoveryRound() ? CycleItem.PoolType.RECOVERY : CycleItem.PoolType.NORMAL;
        CycleItem cycleItem = new CycleItem(
                mCycleConfig.getCycleId(),
                item.getId(),
                poolType,
                mCycleIteration,
                CycleItem.Response.LUPA,
                responseTimeMs,
                1
        );
        mRecordedCycleItems.add(cycleItem);

        // Add to recovery pool for the next iteration
        mCurrentLupaList.add(item);

        mCurrentIndex++;
        checkEndOfPass();
        return mIsComplete;
    }

    private void checkEndOfPass() {
        if (mCurrentIndex >= mCurrentPassItems.size()) {
            if (mCurrentLupaList.isEmpty()) {
                // All items in this pass were remembered!
                mIsComplete = true;
                mCycleConfig.setStatus(LearningCycle.Status.COMPLETED);
                mCycleConfig.setCompletedAtEpoch(System.currentTimeMillis() / 1000L);
            } else {
                // Recovery round needed: strictly repeat only the forgotten items
                mCurrentPassItems = new ArrayList<>(mCurrentLupaList);
                mCurrentLupaList.clear();
                mCycleIteration++;
                mCurrentIndex = 0;
            }
        }
    }
}

File: ios/Kotoba/Engines/IngatLupaEngine.swift
import Foundation

public final class IngatLupaEngine: ObservableObject {
    public let initialPool: [LearningObject]

    @Published public private(set) var currentPassItems: [LearningObject]
    @Published public private(set) var currentLupaList: [LearningObject]
    @Published public private(set) var allResolvedItems: [LearningObject]

    @Published public private(set) var currentIndex: Int = 0
    @Published public private(set) var cycleIteration: Int = 1
    @Published public private(set) var isComplete: Bool = false
    @Published public private(set) var totalIngatCount: Int = 0
    @Published public private(set) var totalLupaCount: Int = 0

    public init(pool: [LearningObject]) {
        self.initialPool = pool
        self.currentPassItems = pool
        self.currentLupaList = []
        self.allResolvedItems = []
        self.isComplete = pool.isEmpty
    }

    public var isRecoveryRound: Bool {
        return cycleIteration > 1
    }

    public var currentItem: LearningObject? {
        guard !isComplete, currentIndex < currentPassItems.count else { return nil }
        return currentPassItems[currentIndex]
    }

    public var currentPassTotal: Int {
        return currentPassItems.count
    }

    public var initialPoolSize: Int {
        return initialPool.count
    }

    public var pendingLupaCount: Int {
        return currentLupaList.count
    }

    public var accuracyPercentage: Double {
        let total = totalIngatCount + totalLupaCount
        guard total > 0 else { return 100.0 }
        return (Double(totalIngatCount) / Double(total)) * 100.0
    }

    @discardableResult
    public func markIngat(responseTimeMs: Int = 1000) -> Bool {
        guard !isComplete, let item = currentItem else { return true }

        totalIngatCount += 1
        if !allResolvedItems.contains(where: { $0.id == item.id }) {
            allResolvedItems.append(item)
        }

        currentIndex += 1
        checkEndOfPass()
        return isComplete
    }

    @discardableResult
    public func markLupa(responseTimeMs: Int = 1000) -> Bool {
        guard !isComplete, let item = currentItem else { return true }

        totalLupaCount += 1
        currentLupaList.append(item)

        currentIndex += 1
        checkEndOfPass()
        return isComplete
    }

    private func checkEndOfPass() {
        if currentIndex >= currentPassItems.count {
            if currentLupaList.isEmpty {
                isComplete = true
            } else {
                currentPassItems = currentLupaList
                currentLupaList = []
                cycleIteration += 1
                currentIndex = 0
            }
        }
    }

    /// Navigate back one card within the current pass.
    /// Silent no-op when already at the first card.
    /// Does NOT undo any committed Ingat/Lupa judgment.
    public func goBack() {
        guard currentIndex > 0 else { return }
        currentIndex -= 1
    }

    public func restart() {
        currentPassItems = initialPool
        currentLupaList = []
        allResolvedItems = []
        currentIndex = 0
        cycleIteration = 1
        totalIngatCount = 0
        totalLupaCount = 0
        isComplete = initialPool.isEmpty
    }
}
